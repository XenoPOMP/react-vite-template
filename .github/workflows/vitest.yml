name: 'Testing'
on:
  push:
  pull_request:

jobs:
  test:
    runs-on: ubuntu-latest

    permissions:
      # Required to checkout the code
      contents: read
      # Required to put a comment into the pull-request
      pull-requests: write

    steps:
      - uses: actions/checkout@v2
      - name: 'Install Node'
        uses: actions/setup-node@v2
        with:
          node-version: '16.x'
      - name: 'Install Deps'
        run: npm i yarn -g && yarn
      - name: 'Test'
        run: yarn coverage
      - name: 'Report Coverage'
        if: always() # Also generate the report if tests are failing
        uses: davelosert/vitest-coverage-report-action@v2
