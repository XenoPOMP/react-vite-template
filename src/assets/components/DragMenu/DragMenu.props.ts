import { ReactChildrenProps } from '../../interfaces/ReactChildren.props';
import { HTMLAttributes } from 'react';

export interface DragMenuProps extends ReactChildrenProps {
  classNames?: {
    constraint?: string,
    draggable?: string,
  };
  axis: 'x' | 'y';
}
