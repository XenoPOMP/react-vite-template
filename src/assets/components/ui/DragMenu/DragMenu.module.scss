.dragElement {
  cursor: grab;

  &:active {
    cursor: grabbing;
  }
}