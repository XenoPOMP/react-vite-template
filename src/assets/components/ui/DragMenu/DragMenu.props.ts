export interface DragMenuProps {
  classNames?: {
    constraint?: string;
    draggable?: string;
  };
  axis: 'x' | 'y';
}
