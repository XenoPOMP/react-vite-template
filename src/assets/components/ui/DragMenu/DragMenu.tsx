import cn from 'classnames';
import { motion } from 'framer-motion';
import { FC, PropsWithChildren, useRef } from 'react';

import styles from './DragMenu.module.scss';
import type { DragMenuProps } from './DragMenu.props';

/**
 * Draggable component.
 *
 * @param children
 * @param classNames
 * @param {"x"|"y"} axis
 * @constructor
 */
const DragMenu: FC<PropsWithChildren<DragMenuProps>> = ({
  children,
  classNames,
  axis,
}) => {
  const constraintRef = useRef(null);

  return (
    <motion.div
      ref={constraintRef}
      className={cn(classNames?.constraint, styles.dragElement)}
    >
      <motion.div
        drag={axis}
        dragConstraints={constraintRef}
        dragElastic={0.2}
        className={classNames?.draggable}
      >
        {children}
      </motion.div>
    </motion.div>
  );
};

export default DragMenu;
