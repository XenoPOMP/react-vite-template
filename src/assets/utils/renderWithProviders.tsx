import { RenderResult, render } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider as ReduxProvider } from 'react-redux';
import { Route, BrowserRouter as Router, Routes } from 'react-router-dom';

import store from '@redux/index';

/**
 * Render options for Vitest DOM wrapping
 */
export interface RenderOptions {
  useRedux?: boolean;
  useRouter?: boolean;
  useQuery?: boolean;
}

/**
 * Type for UI wrapper.
 */
type Wrapper = (ui: JSX.Element) => JSX.Element;

/**
 * Wraps UI in Redux store.
 * @param {JSX.Element} ui     wrapping UI
 * @returns {JSX.Element}      wrapped UI
 */
const wrapInRedux: Wrapper = ui => {
  return <ReduxProvider store={store}>{ui}</ReduxProvider>;
};

/**
 * Wraps UI in Router.
 * @param {JSX.Element} ui     wrapping UI
 * @returns {JSX.Element}      wrapped UI
 */
const wrapInRouter: Wrapper = ui => {
  return (
    <Router>
      <Routes>
        <Route path={'/'} element={ui} />
      </Routes>
    </Router>
  );
};

/**
 * Wraps UI in React Query.
 * @param {JSX.Element} ui     wrapping UI
 * @returns {JSX.Element}      wrapped UI
 */
const wrapInReactQuery: Wrapper = ui => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        refetchOnWindowFocus: true,
        // refetchInterval: 5000
      },
    },
  });

  return <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>;
};

/**
 * Renders JSX wrapped in providers.
 * @param {JSX.Element} ui
 * @param options
 * @returns {RenderResult}
 */
const renderWithProviders = (ui: JSX.Element, options?: RenderOptions) => {
  const renderOptions: RenderOptions = {
    useRedux: false,
    useRouter: false,
    useQuery: false,
    ...options,
  };
  let renderUi: JSX.Element = ui;

  if (renderOptions.useRedux) {
    renderUi = wrapInRedux(renderUi);
  }

  if (renderOptions.useRouter) {
    renderUi = wrapInRouter(renderUi);
  }

  if (renderOptions.useQuery) {
    renderUi = wrapInReactQuery(renderUi);
  }

  return render(renderUi);
};

export default renderWithProviders;
