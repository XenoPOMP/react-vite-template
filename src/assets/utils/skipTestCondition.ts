import useEnv, { TestingMode } from '@hooks/useEnv';

const { TESTING_MODE } = useEnv();

/**
 * Determines skip test condition.
 * @param {TestingMode} mode       target mode
 * @returns {boolean}              false if current mode isn`t the same as target one
 *                                 or current mode is FULLSTACK
 */
const skipTestCondition = (mode: TestingMode): boolean => {
  if (TESTING_MODE === 'FULLSTACK') {
    return false;
  }

  return TESTING_MODE !== mode;
};

export { skipTestCondition as default };
