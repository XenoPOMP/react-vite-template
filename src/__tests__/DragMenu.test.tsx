import { describe, expect, test } from 'vitest';

import DragMenu from '@ui/DragMenu/DragMenu';

import renderWithProviders from '@utils/renderWithProviders';
import skipTestCondition from '@utils/skipTestCondition';

describe.skipIf(skipTestCondition('FRONTEND'))('Draggable menu', () => {
  test('Match snapshot', () => {
    expect(
      renderWithProviders(
        <DragMenu axis={'x'}>
          <div>Drag me</div>
        </DragMenu>,
      ),
    ).toMatchSnapshot();

    expect(
      renderWithProviders(
        <DragMenu
          axis={'x'}
          classNames={{
            constraint: 'parent class',
          }}
        >
          <div>Drag me</div>
        </DragMenu>,
      ),
    ).toMatchSnapshot();

    expect(
      renderWithProviders(
        <DragMenu
          axis={'x'}
          classNames={{
            constraint: 'parent class',
            draggable: 'child class',
          }}
        >
          <div>Drag me</div>
        </DragMenu>,
      ),
    ).toMatchSnapshot();
  });
});
