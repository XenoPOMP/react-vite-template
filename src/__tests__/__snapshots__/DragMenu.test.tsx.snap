// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Draggable menu > Match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_dragElement_960780"
      >
        <div
          draggable="false"
          style="user-select: none;"
        >
          <div>
            Drag me
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_dragElement_960780"
    >
      <div
        draggable="false"
        style="user-select: none;"
      >
        <div>
          Drag me
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Draggable menu > Match snapshot 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_dragElement_960780"
      >
        <div
          draggable="false"
          style="user-select: none;"
        >
          <div>
            Drag me
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="parent class _dragElement_960780"
      >
        <div
          draggable="false"
          style="user-select: none;"
        >
          <div>
            Drag me
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="parent class _dragElement_960780"
    >
      <div
        draggable="false"
        style="user-select: none;"
      >
        <div>
          Drag me
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Draggable menu > Match snapshot 3`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_dragElement_960780"
      >
        <div
          draggable="false"
          style="user-select: none;"
        >
          <div>
            Drag me
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="parent class _dragElement_960780"
      >
        <div
          draggable="false"
          style="user-select: none;"
        >
          <div>
            Drag me
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="parent class _dragElement_960780"
      >
        <div
          class="child class"
          draggable="false"
          style="user-select: none;"
        >
          <div>
            Drag me
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="parent class _dragElement_960780"
    >
      <div
        class="child class"
        draggable="false"
        style="user-select: none;"
      >
        <div>
          Drag me
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
